packages/config-loader/src/lib/transform/include.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import yaml from 'yaml';
+import { extname, dirname, resolve as resolvePath } from 'path';
+import { JsonObject, JsonValue } from '@backstage/config';
+import { isObject } from './utils';
+import { TransformFunc, EnvFunc, ReadFileFunc } from './types';
+
+// Parsers for each type of included file
+const includeFileParser: {
+  [ext in string]: (content: string) => Promise<JsonObject>;
+} = {
+  '.json': async content => JSON.parse(content),
+  '.yaml': async content => yaml.parse(content),
+  '.yml': async content => yaml.parse(content),
+};
+
+/**
+ * Transforms a secret description into the actual secret value.
+ */
+export function createIncludeTransform(
+  env: EnvFunc,
+  readFile: ReadFileFunc,
+): TransformFunc {
+  return async (input: JsonValue, baseDir: string) => {
+    if (!isObject(input)) {
+      return { applied: false };
+    }
+    // Check if there's any key that starts with a '$', in that case we treat
+    // this entire object as a secret.
+    const [secretKey] = Object.keys(input).filter(key => key.startsWith('$'));
+    if (secretKey) {
+      if (Object.keys(input).length !== 1) {
+        throw new Error(
+          `include key ${secretKey} should not have adjacent keys`,
+        );
+      }
+    } else {
+      return { applied: false };
+    }
+
+    const secretValue = input[secretKey];
+    if (typeof secretValue !== 'string') {
+      throw new Error(`${secretKey} include value is not a string`);
+    }
+
+    switch (secretKey) {
+      case '$file':
+        try {
+          const value = await readFile(resolvePath(baseDir, secretValue));
+          return { applied: true, value };
+        } catch (error) {
+          throw new Error(`failed to read file ${secretValue}, ${error}`);
```

Comment:
just randomly noting here - all these have a lowercase initial character

Reply:
Yeah went for structuring the error messages with `error.message` here as they're all just regular `Error` instances and meant for direct display to users.

Reply:
Sure, I just was under the impression that this would be `Failed` elsewhere.

Reply:
Yeah elsewhere I would've gone for `Failed`, but doing some very controlled throwing here, trying to construct actual sentences rather than just patch all errors together :p